kratos/solving_strategies/schemes/residual_based_bdf_scheme.h

```diff
@@ -0,0 +1,582 @@
+
+    ///@name Protected static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected member Variables
+    ///@{
+
+    struct GeneralVectors
+    {
+        std::vector< Vector > dotun0; /// First derivative
+        std::vector< Vector > dot2un0; /// Second derivative
+    };
+    
+    const std::size_t mOrder; /// The integration order
+    Vector mBDF; /// The BDF coefficients
+    GeneralVectors mVector; /// The structure containing the  derivatives
+
+    ///@}
+    ///@name Protected Operators
+    ///@{
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+    
+    /**
+     * @brief Performing the update of the derivatives
+     * @param rModelPart The model of the problem to solve
+     * @param rDofSet Set of all primary variables
+     * @param A LHS matrix
+     * @param Dx incremental update of primary variables
+     * @param b RHS Vector
+     */
+    inline void UpdateDerivatives(
+        ModelPart& rModelPart,
+        DofsArrayType& rDofSet,
+        TSystemMatrixType& A,
+        TSystemVectorType& Dx,
+        TSystemVectorType& b
+        )
+    {
+        // Updating time derivatives (nodally for efficiency)
+        const int num_nodes = static_cast<int>( rModelPart.Nodes().size() );
+
+        #pragma omp parallel for
+        for(int i = 0;  i< num_nodes; ++i) {
+            auto it_node = rModelPart.Nodes().begin() + i;
+            
+            UpdateFirstDerivative(it_node);
+            UpdateSecondDerivative(it_node);
+        }
+    }
+    
+    /**
+     * @brief Updating first time derivative (velocity)
+     * @param itNode the node interator
+     */
+    
+    virtual inline void UpdateFirstDerivative(NodesArrayType::iterator itNode)
+    {
+        KRATOS_ERROR << "Calling base BDF class" << std::endl;
+    }
+
+    /**
+     * @brief Updating second time derivative (acceleration)
+     * @param itNode the node interator
+     */
+    
+    virtual inline void UpdateSecondDerivative(NodesArrayType::iterator itNode)
+    {
+        KRATOS_ERROR << "Calling base BDF class" << std::endl;
+    }
+    
+    /**
+     * @brief It adds the dynamic LHS contribution of the elements 
+     * \f[ LHS = \frac{d(-RHS)}{d(u_{n0})} = c_0^2\mathbf{M} + c_0 \mathbf{D} + \mathbf{K} \f]
+     * @param LHS_Contribution The dynamic contribution for the LHS
+     * @param D The damping matrix
+     * @param M The mass matrix
+     * @param rCurrentProcessInfo The current process info instance
+     */
+
+    void AddDynamicsToLHS(
+        LocalSystemMatrixType& LHS_Contribution,
+        LocalSystemMatrixType& D,
+        LocalSystemMatrixType& M,
+        ProcessInfo& rCurrentProcessInfo
+        ) override
+    {
+        // Adding mass contribution to the dynamic stiffness
+        if (M.size1() != 0) { // if M matrix declared
+            noalias(LHS_Contribution) += M * std::pow(mBDF[0], 2);
+        }
+
+        // Adding  damping contribution
+        if (D.size1() != 0) { // if D matrix declared
```

Comment:
I am thinking if there is a way (maybe function in the spaces) to check if the dampingmatrix does not contain values.
then you could improve your check

E.g. in StructuralMechanics many elements only resize the dampingmatrix, but they don't fill it.
In this case your check would fail, i.e. more computational effort

Reply:
Ok

Reply:
I couldn't find any method for that in ublas_space

Reply:
In anyc ase if the matrix is empty of zero will have no effect, 0*x = 0, so it is OK